abide/src/main/scala/scala/tools/abide/traversal/ScopingRule.scala

```diff
@@ -1,40 +1,1113 @@
 package scala.tools.abide.traversal
 
 import scala.reflect.internal.traversal._
+import scala.collection.{ mutable, immutable }
+import scala.annotation.tailrec
 
 /**
  * ScopingRule
  *
- * TraversalRule subtrait that provides helper methods to manage scoping during traversal. Then [[enter]] method
- * will push the current tree to the scope and register a leaver method that will pop it once we leave that tree.
+ * TraversalRule subtrait that provides scoping from [[ScopingRuleTraversal]] with simple warning accumulation
+ * from [[WarningRuleTraversal]], meaning warnings are accumulated given scoping information and local context.
  *
- * As in [[WarningRule]], warnings are determined given local (and scoping) context in a single pass (no
- * validation/invalidation mechanism).
+ * @see [[ScopingRuleTraversal]] for scala-style scoping aspects
+ * @see [[WarningRuleTraversal]] for warning accumulation
  */
-trait ScopingRule extends TraversalRule with ScopingTraversal {
+trait ScopingRule extends ScopingRuleTraversal with WarningRuleTraversal {
   import context.universe._
 
-  /** Scoping type (eg. method symbol, class symbol, etc.) */
-  type Owner
+  def emptyState = State(NoContext, Nil)
+  case class State(context: Context, warnings: List[Warning]) extends ScopingState with WarningState {
+    def withContext(newContext: Context): State = State(newContext, warnings)
+    def nok(warning: Warning): State = State(context, warning :: warnings)
+  }
+}
+
+/**
+ * ScopingRuleTraversal
+ *
+ * TraversalRule subtrait that provides scala-style symbol scoping and lets rule writers lookup symbols
+ * given names and filters. To ensure the scoping mechanism resembles the one implemented in the scala
+ * compiler, the scope building is handled automatically by the ScopingRuleTraversal trait.
+ *
+ * The trait provides a [[lookup]] method to find the first symbol in scope given `name` and `qualifies`
+ * filter function. Certain parts of scala scoping were ommited as not every tree is useful for scope
+ * lookup. The scoping context is complete for the following ASTs in scala.reflect.internal.Trees:
```

Comment:
Well, unfortunately the list of ASTs that are not handled is also quite hefty. Basically, the difference should be that all ASTs that are not in the list won't have associated scopes / belong to the scope.

Given the complexity of scoping in the compiler implementation, I can't even guarantee my implementation is really equivalent, but differences should be rare for the trees I listed.